core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1608,16 +1629,20 @@ public long newRequestId() {
      * Called once the channel is closed for instance due to a disconnect or a closed socket etc.
      */
     protected final void onChannelClosed(Channel channel) {
-        final Optional<Long> first = pendingHandshakes.entrySet().stream()
-            .filter((entry) -> entry.getValue().channel == channel).map((e) -> e.getKey()).findFirst();
-        if (first.isPresent()) {
-            final Long requestId = first.get();
-            final HandshakeResponseHandler handler = pendingHandshakes.remove(requestId);
-            if (handler != null) {
-                // there might be a race removing this or this method might be called twice concurrently depending on how
-                // the channel is closed ie. due to connection reset or broken pipes
-                handler.handleException(new TransportException("connection reset"));
+        try {
+            final Optional<Long> first = pendingHandshakes.entrySet().stream()
+                .filter((entry) -> entry.getValue().channel == channel).map((e) -> e.getKey()).findFirst();
+            if (first.isPresent()) {
+                final Long requestId = first.get();
+                final HandshakeResponseHandler handler = pendingHandshakes.remove(requestId);
+                if (handler != null) {
+                    // there might be a race removing this or this method might be called twice concurrently depending on how
+                    // the channel is closed ie. due to connection reset or broken pipes
+                    handler.handleException(new TransportException("connection reset"));
+                }
             }
+        } finally {
+            disconnectFromNodeChannel(channel, "channel closed");
```

Comment:
question - why did you add this? it currently has a nasty side effect of spawning extra "disconnect" generic threads when we are already disconnecting. Say we have an issue on one channel, we find the relevant NodeChannels and close all other channels in it. Those will end up here, spawning another disconnectFromNodeChannel call, which will be a no-op (as we are cleaning it up), but we're creating all kind of concurrency issues. I'm wondering why we need this.